File: android/share/src/main/java/com/dreampany/share/data/source/repository/nearby/NearbyRepository.java
package com.dreampany.share.data.source.repository.nearby;

import android.content.Context;

import com.annimon.stream.Stream;
import com.annimon.stream.function.Consumer;
import com.dreampany.framework.data.source.repository.Repository;
import com.dreampany.framework.misc.ResponseMapper;
import com.dreampany.framework.misc.RxMapper;
import com.dreampany.framework.util.AndroidUtil;
import com.dreampany.framework.util.DataUtil;
import com.dreampany.nearby.NearbyApi;
import com.dreampany.nearby.model.Peer;
import com.dreampany.share.data.model.User;
import com.google.common.collect.Sets;

import java.util.Set;

import javax.inject.Inject;
import javax.inject.Singleton;

/**
 * Created by Hawladar Roman on 9/4/2018.
 * BJIT Group
 * hawladar.roman@bjitgroup.com
 */
@Singleton
public class NearbyRepository extends Repository<Long, User> implements NearbyApi.Callback {

    public interface UserCallback {
        void onJoin(User user);

        void onUpdate(User user);

        void onLeave(User user);
    }

    public interface DiscoverCallback {

    }

    private final Context context;
    private final NearbyApi nearby;

    private long serviceId;
    private User owner;

    private int clients;
    private final Set<UserCallback> userCallbacks;
    private final Set<DiscoverCallback> discoverCallbacks;

    @Inject
    NearbyRepository(Context context,
                     RxMapper rx,
                     ResponseMapper rm,
                     NearbyApi nearby) {
        super(rx, rm);
        this.context = context;
        this.nearby = nearby;

        if (owner == null) {
            serviceId = DataUtil.getSha512(AndroidUtil.getApplicationId(context));
            owner = buildOwner();
        }

        nearby.register(this);
        userCallbacks = Sets.newConcurrentHashSet();
        discoverCallbacks = Sets.newConcurrentHashSet();
    }

    @Override
    public void onPeer(Peer peer, Peer.State state) {
        switch (state) {
            case LIVE:
                Stream.of(userCallbacks).forEach(new Consumer<UserCallback>() {
                    @Override
                    public void accept(UserCallback callback) {
                        //callback.onJoin();
                    }
                });
                break;
        }
    }

    @Override
    public void onData(Peer peer, byte[] data) {

    }

    @Override
    public void onStatus(long payloadId, NearbyApi.PayloadState state, long totalBytes, long bytesTransferred) {

    }

    public void register(UserCallback callback) {
        userCallbacks.add(callback);
        start();
    }

    public void register(DiscoverCallback callback) {
        discoverCallbacks.add(callback);
        start();
    }

    public void unregister(UserCallback callback) {
        userCallbacks.remove(callback);
        stop();
    }

    public void unregister(DiscoverCallback callback) {
        discoverCallbacks.remove(callback);
        stop();
    }

    private void start() {
        if (!userCallbacks.isEmpty() || !discoverCallbacks.isEmpty()) {
            nearby.init(context, serviceId, owner.getId());
            nearby.start();
        }
/*        if (clients <= 0) {
            nearby.start();
        }
        clients++;*/
    }

    private void stop() {
        if (userCallbacks.isEmpty() && discoverCallbacks.isEmpty()) {
            nearby.stop();
        }
/*        clients--;
        if (clients <= 0) {
            nearby.stop();
        }*/
    }


    private User buildOwner() {
        User user = new User();
        user.setId(DataUtil.getSha512(AndroidUtil.getAndroidId(context)));
        user.setName(AndroidUtil.getDeviceName());
        return user;
    }
}

File: android/cast/src/main/kotlin/com/dreampany/cast/data/source/repository/NearbyRepository.kt
package com.dreampany.cast.data.source.repository

import android.content.Context
import com.dreampany.cast.data.model.User
import com.dreampany.frame.data.source.repository.RepositoryKt
import com.dreampany.frame.misc.ResponseMapper
import com.dreampany.frame.misc.RxMapper
import com.dreampany.frame.util.AndroidUtil
import com.dreampany.frame.util.DataUtil
import com.dreampany.frame.util.TimeUtil
import com.dreampany.nearby.NearbyApi
import com.dreampany.nearby.model.Peer
import timber.log.Timber
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Created by Roman-372 on 7/16/2019
 * Copyright (c) 2019 bjit. All rights reserved.
 * hawladar.roman@bjitgroup.com
 * Last modified $file.lastModified
 */
@Singleton
class NearbyRepository @Inject constructor(
    private val context: Context,
    rx: RxMapper,
    rm: ResponseMapper,
    private val nearby: NearbyApi
) : RepositoryKt<String, User>(rx, rm), NearbyApi.Callback {

    interface UserCallback {
        fun onJoin(user: User)

        fun onUpdate(user: User)

        fun onLeave(user: User)
    }

    private var serviceId: Long = 0
    private val owner: User
    private val ownerId: Long

    private val userCallbacks: MutableSet<UserCallback>

    init {
        serviceId = DataUtil.getSha512(AndroidUtil.getPackageName(context))
        owner = buildOwner()
        ownerId = DataUtil.getSha512(owner.id)

        nearby.register(this)
        userCallbacks = mutableSetOf()
    }

    override fun onPeer(peer: Peer, state: Peer.State) {
        Timber.v("Peer (%d) - State(%s)", peer.id, state.toString())
    }

    override fun onData(peer: Peer, data: ByteArray) {
    }

    override fun onStatus(
        payloadId: Long,
        state: NearbyApi.PayloadState,
        totalBytes: Long,
        bytesTransferred: Long
    ) {
    }

    fun register(callback: UserCallback) {
        userCallbacks.add(callback)
        start()
    }

    fun unregister(callback: UserCallback) {
        userCallbacks.remove(callback)
        stop()
    }

    private fun start() {
        if (!userCallbacks.isEmpty() /*|| !discoverCallbacks.isEmpty()*/) {
            nearby.init(context, serviceId, ownerId)
            nearby.start()
        }
        /*        if (clients <= 0) {
            nearby.start();
        }
        clients++;*/
    }

    private fun stop() {
        if (userCallbacks.isEmpty() /*&& discoverCallbacks.isEmpty()*/) {
            nearby.stop()
        }
        /*        clients--;
        if (clients <= 0) {
            nearby.stop();
        }*/
    }

    private fun buildOwner(): User {
        val id = AndroidUtil.getAndroidId(context)
        val time = TimeUtil.currentTime()
        val name = AndroidUtil.getDeviceName()
        val user = User(id, time)
        user.name = name
        return user
    }
}